controllers/cloudstackfailuredomain_controller.go

```diff
@@ -130,21 +140,222 @@ func (r *CloudStackFailureDomainReconciliationRunner) ClearMachines() (ctrl.Resu
+		for _, ref := range machine.OwnerReferences {
+			if ref.Kind == "EtcdadmCluster" {
+				etcdMachineFound = true
+				etcdMachine = machine
+			}
+		}
+	}
+	if !etcdMachineFound {
+		r.Log.Info("Clear machine: no etcd VM found.")
+		return ctrl.Result{}, nil
+	}
+	for _, ref := range etcdMachine.OwnerReferences {
+		if ref.Kind == "EtcdadmCluster" {
+			r.Log.Info("Clear machine: getting etcdadmcluster ...")
+			etcdadmCluster := &unstructured.Unstructured{}
+			etcdadmCluster.SetGroupVersionKind(schema.FromAPIVersionAndKind(ref.APIVersion, ref.Kind))
+			if err := r.K8sClient.Get(r.RequestCtx, client.ObjectKey{Namespace: etcdMachine.Namespace, Name: ref.Name}, etcdadmCluster); err != nil {
+				return ctrl.Result{}, err
+			}
+			r.Log.Info("Clear machine: retrieved etcdadmcluster.")
+			csMachineTemplateName, ok := etcdMachine.Annotations["cluster.x-k8s.io/cloned-from-name"]
+			if !ok {
+				result, err := r.RequeueWithMessage("annotations['cluster.x-k8s.io/cloned-from-name'] not found.", "cloudstackmachine", etcdMachine.Name)
+				return result, err
+			}
+			r.Log.Info(fmt.Sprintf("Clear machine: machine template name %s decided", csMachineTemplateName))
+			csMachineTemplateNameInEtcdadmCluster, found, err := unstructured.NestedString(etcdadmCluster.Object, "spec", "infrastructureTemplate", "name")
+			if err != nil || !found {
+				return ctrl.Result{}, errors.New("etcdadmcluster spec.infrastructureTemplate.name not found or not string")
+			}
+			if csMachineTemplateName != csMachineTemplateNameInEtcdadmCluster {
+				return ctrl.Result{}, errors.Errorf("cloudstackmachinetemplate in machine %s and etcdadmcluster %s are different", etcdMachine.Name, ref.Name)
+			}
+			csMachineTemplate := &infrav1.CloudStackMachineTemplate{}
+			if err := r.K8sClient.Get(r.RequestCtx, client.ObjectKey{Namespace: etcdMachine.Namespace, Name: csMachineTemplateName}, csMachineTemplate); err != nil {
+				return ctrl.Result{}, err
+			}
+			r.Log.Info(fmt.Sprintf("Clear machine: machine template name %s retrieved", csMachineTemplateName))
+			csMachineTemplate.Name = fmt.Sprintf("%s-template-%d", ref.Name, time.Now().UnixNano()/int64(time.Millisecond))
+			csMachineTemplate.SetAnnotations(map[string]string{})
+			csMachineTemplate.SetResourceVersion("")
+			if err := r.K8sClient.Create(r.RequestCtx, csMachineTemplate); err != nil {
+				return ctrl.Result{}, err
+			}
+			r.Log.Info(fmt.Sprintf("Clear machine: new machine template name %s created", csMachineTemplate.Name))
+
+			if _, err := controllerutil.CreateOrPatch(r.RequestCtx, r.K8sClient, etcdadmCluster,
+				func() error {
+					return unstructured.SetNestedField(
+						etcdadmCluster.Object,
+						csMachineTemplate.Name,
+						"spec", "infrastructureTemplate", "name",
+					)
+				}); err != nil {
+				return ctrl.Result{}, err
+			}
+			r.Log.Info(fmt.Sprintf("Clear machine: etcdadmcluster infrastructureTemplate cloudstackmachinetemplate name %s patched", csMachineTemplate.Name))
```

Comment:
could you move this into a helper function to make it more readable? I also wonder if we could reuse some of this code between the different deployment triggers